### DISCUSSION TITLE
how to programatically prevent revalidation of pages?

### DISCUSSION THREAD

**Discussion by tomaszrondio** [vercel/next.js#31345] [Help] [answered]
use case: 
getStaticProps calls API. API returns error. I want to prevent that single page (part of dynamic route) to be revalidated. 
Preferably without throwing and error. I just need next to skip revalidating when API is not responding etc.

**Comment by balazsorban44** [comment]
When an error occurs in `getStaticProps`, Next.js won't cache the result and will redirect you to the error page. Next time you visit the same page, you will keep ending up on the error page, until the error isn't fixed.

If your API returns an error and you opt-out revalidation, that means the page will be "stuck" in a broken state forever/until the next build.

Are you sure that is what you need?

**Reply by tomaszrondio** [reply]
U got me wrong. I have a valid page prerendered. Then something happens to backend and it errors. I want to keep working version cached and do not revalidate untill backend is fine again.

**Reply by balazsorban44** [reply]
I see!

If you tell `getStaticProps` to `revalidate` after a certain time, it **will** try to revalidate and it doesn't have control over your API's success. (FYI, it will try to fetch the resource a few times before bailing out and throwing an error.)

I think what you could do is that in your API, build in error handling that will return the old data in case there was an error.

From a user perspective, if you wanted to revalidate the page, I would have assumed the user wanted to know that an error happened, so you might want to show the user the reason the content is potentially outdated. No?

**Reply by tomaszrondio** [reply]
that is the point that sometimes API service can be down. Due to Cloudflare issues, server issues or DDoS attack. In these cases I want to tell next "hey, don't revalidate that page, leave it as it is" bcs otherwise I end up with blank page.

**Reply by balazsorban44** [reply]
Such an event hopefully isn't the standard, and the outage would only last for a short time. In that case, I would recommend to make the user aware of the outage in some other way (like a status page maybe?)

When the API stabilized itself, the page will be regenerated as expected and everything goes back to normal.

Setting up proper monitoring/logging/testing of such events/server issues might also help avoid the problem in the first place.

What do you think?

**Reply by tomaszrondio** [reply]
the problem is that in my world it is a everyday bread. We work for news outlets that are DDoS attacked for a week straight (from time to time. Usually when they publish something against government etc.). So the idea is to have another layer of protection + it is easier to scale next app than whole setup including backend.

**Reply by balazsorban44** [reply]
Next.js dedupes requests when revalidating a page https://github.com/vercel/next.js/pull/30125, so I think your API would never be hit more than once.

**Reply by tomaszrondio** [reply]
But my problem persists if API won't respond that one time. Right?

**Reply by balazsorban44** [reply]
yes, but it won't be because of DDoS. then all my previous suggestions apply.

**Reply by tomaszrondio** [reply]
It does but it will be still very nice to have a way to control revalidation process and stop it if needed.

**Reply by balazsorban44** [reply]
There is little use-case in my personal opinion. Errors should be handled/prevented not ignored.

**Reply by tomaszrondio** [reply]
exactly! So instead of ignoring ("it will work in 5 minutes") I want to prevent them by keeping static files filled with content.

**Reply by balazsorban44** [reply]
When you show old content, you mask the error.

When you set `revalidate` to a value, you indicate Next.js when it should fetch new content. If that fails for some reason, the user should be made aware of why they are receiving potentially outdated content.

If you are fine with showing that old data anyway, you can probably just tweak the `revalidate` value to be higher instead, thus reducing the possibility when an error is returned.

**Reply by sam-jg** [reply]
I am also looking for a similar solution to @tomaszrondio. In my case we are using `getStaticProps` to fetch instagram and blog posts to display on a website.

@balazsorban44 I agree that errors should be dealt with but don't think it really applies for all content. It makes sense for content specific to a user that they can update regularly, but for something like public blog posts it is much better to display something (even if it is out of date) than nothing at all.

I was thinking that the existing props data could be included in the context parameter and then optionally returned with some error handling like this.
```
export const getStaticProps = async (context) => {
  try {
    const res = await axios.get(
      `https://graph.instagram.com/v12.0/${INSTAGRAM_ACCOUNT_ID}/media`,
      {
        params: {
          access_token: INSTAGRAM_ACCESS_TOKEN,
          fields: 'id,media_type,media_url,permalink,thumbnail_url',
        },
      },
    );
    return {
      props: {
        instagramPosts: res.data.data,
      },
      revalidate: PAGE_REVALIDATE_TIME,
    };
  } catch (error) {
    return {
      props:  context.props,
      revalidate: PAGE_REVALIDATE_TIME,
    };
  }
};
```

**Reply by balazsorban44** [reply]
I still don't think it is Next.js's responsibility to mask the errors, but to answer the original question, I do not think that it is currently supported to prevent revalidation of a page when the revalidation is ought to happen.

If you still want to return stale data when your API fails, maybe the `stale-if-error` Cache-Control header could be considered on your API: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#:~:text=will%20revalidate%20normally.-,stale%2Dif%2Derror,-The%20stale%2Dif


**My bad, correction on my answers**: See https://github.com/vercel/next.js/discussions/31345#discussioncomment-1650538

**Reply by sam-jg** [reply]
I don't really see how a change like the one I suggested is moving any responsibility to Next.js. If I understand Tim's answer correctly, then Next.js has already taken the responsibility to mask all uncaught errors by keeping the previous page and provided no way to show that the page is stale. Including the existing props in the context means that we could return them along with some additional value indicating that there has been some error and the data is not up to date.

```
} catch (error) {
    return {
      props:  {
          ...context.props,
          error,
          stale: true
      },
      revalidate: PAGE_REVALIDATE_TIME,
    };
}
```

**Reply by balazsorban44** [reply]
it was my bad, Next.js already works as `stale-if-error`

**Comment by timneutkens** [comment, marked as answer]
> getStaticProps calls API. API returns error. I want to prevent that single page (part of dynamic route) to be revalidated.
Preferably without throwing and error. I just need next to skip revalidating when API is not responding etc.

The behavior of `revalidate` is to keep the previously generated page if an error is thrown.

```js
export async function getStaticProps() {
   const res = await fetch('https://examples.com/my-api')
   if(res.status === 500) {
      throw new Error('Not revalidated')
   }
   return {
    revalidate: 1,
    props: {}
  }
}
```

**Reply by tomaszrondio** [reply]
I knew about it but wanted to avoid throwing errors. Anyway now I know that there is no other option than throwing an error. Thank You.

**Reply by sam-jg** [reply]
@timneutkens is this the behaviour in both a development and production environment? I am just getting the unhandled error message at the moment.

**Reply by gabrielreisn** [reply]
> The behavior of revalidate is to keep the previously generated page if an error is thrown.

Is that still the same behavior on app router?
